### PR TITLE
Multiple linear regression 2

### PR DESCRIPTION
**Pull request by MuhamedEssam**
Please have alook on new updates for these 2 files

### CODE REVIEW

File: client/tinychain/ml/mlr.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        d_weights = 2/inputs.shape[0]*(einsum("ij,ki->kj", weights, einsum("ij,ki->kj", inputs, inputs.transpose()))-
+                                                      einsum("ij,ki->kj", labels, inputs.transpose()))        
+        return d_weights
+
+
+    def update_params(self ,weights, d_weights): 
+        new_weights = weights - self.learningrate * d_weights
+        return new_weights
+
+    def fit(self, X_train, y_train,iteration):
```
Comment (haydnv) [inline client/tinychain/ml/mlr.py:52]: Could you use the existing `GradientDescent` optimizer for this instead?

File: client/tinychain/ml/mlr.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+from tinychain.ml import Layer, NeuralNet
+from tinychain.ref import After
+import tinychain as tc 
+import numpy as np
+
+
+class MultipleLinearRegression :
+    
+    @classmethod
+    def load (self,learningrate):
```
Comment (haydnv) [inline client/tinychain/ml/mlr.py:13]: A `classmethod` doesn't have a `self` parameter, also I think this `load` method is basically redundant with `MultipleLinearRegression.__init__`

File: client/tinychain/ml/mlr.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+
+class MultipleLinearRegression :
+    
+    @classmethod
+    def load (self,learningrate): 
+        return MultipleLinearRegression([learningrate])
+    
+    @property
+    def learningrate(self):
```
Comment (haydnv) [inline client/tinychain/ml/mlr.py:17]: `MultipleLinearRegression` doesn't subclass any TinyChain native class (like `Tuple`) so you don't need to access properties this way

File: tests/client/test_ml.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        cxt.mlr = MultipleLinearRegression.load(cxt.learningrate)       
+        cxt.weights= cxt.mlr.fit([cxt.X_train,cxt.y_train,cxt.iteration])
+        cxt.preds=cxt.mlr.predict(cxt.X_test ,learning=False)
+        
+        cxt.score=cxt.mlr.R2_Score(cxt.X_test,cxt.y_test)
+        self.assertGreaterEqual(cxt.score,0.9)
+               
+        
+        response=self.host.post(ENDPOINT,cxt)
+        print(response)
```
Comment (haydnv) [inline tests/client/test_ml.py:72]: you should assert that the regression model fits the data to some specific degree of accuracy

File: tests/client/test_ml.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import math
+
+import numpy as np
+import tinychain as tc
+import unittest
+from tinychain.collection.tensor import einsum, Dense , Tensor
+
+
+from testutils import ClientTest
+import tinychain.ml.mlr import MultipleLinearRegression
```
Comment (haydnv) [inline tests/client/test_ml.py:13]: `from tinychain.ml.mlr import ...`

File: tests/client/test_ml.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        
+        cxt.learningrate = tc.Number(0.1)
+        cxt.iteration= tc.Number(400)
+        cxt.mlr = MultipleLinearRegression.load(cxt.learningrate)       
+        cxt.weights= cxt.mlr.fit([cxt.X_train,cxt.y_train,cxt.iteration])
+        cxt.preds=cxt.mlr.predict(cxt.X_test ,learning=False)
+        
+        cxt.score=cxt.mlr.R2_Score(cxt.X_test,cxt.y_test)
+        self.assertGreaterEqual(cxt.score,0.9)
+
```
Comment (haydnv) [inline tests/client/test_ml.py:69]: when I try to run this test I get an error on this line:
```
$ python3 tests/client/test_ml.py 
Traceback (most recent call last):
  File "/home/haydn/Documents/tinychain/tests/client/test_ml.py", line 13, in <module>
    from tinychain.ml.mlr import MultipleLinearRegression 
  File "/home/haydn/Documents/tinychain/client/tinychain/ml/mlr.py", line 69
    ​
    ^
SyntaxError: invalid non-printable character U+200B
```

**Review by haydnv — CHANGES_REQUESTED** [review]

File: client/tinychain/ml/mlr.py (hunk 1 above)
Comment (MuhamedEssam) [reply client/tinychain/ml/mlr.py:52]: i will search for it i guess maybe its in nn.py ?

File: tests/client/test_ml.py (hunk 6 above)
Comment (MuhamedEssam) [reply tests/client/test_ml.py:69]: i will check it again

File: client/tinychain/ml/mlr.py (hunk 1 above)
Comment (haydnv) [reply client/tinychain/ml/mlr.py:52]: It's in the `optimizer` package: https://github.com/haydnv/tinychain/blob/main/client/tinychain/ml/optimizer.py

File: tests/client/test_ml.py (hunk 4 above)
Comment (haydnv) [reply tests/client/test_ml.py:72]: yes but the test should assert that, e.g. `self.assertGreater(accuracy > 0.9)`

more info: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertGreater

**Comment by haydnv** [comment]
Superceded by #115

**Comment by haydnv** [comment]
Superceded by #115

**Pull request closed without merging**